File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.util.*;

public class Main {
    static ArrayList<Integer>[] edge = new ArrayList[100001];
    static int[] parent = new int[100001];
    static boolean[] marked = new boolean[100001];
    
    public static void dfs(int n)
    {
        marked[n] = true;
        for(int x : edge[n])
        {
            if(!marked[x])
            {
                parent[x] = n;
                dfs(x);
            }
        }
    }
    public static void main(String args[]) {
      
      Scanner s = new Scanner(System.in);
      int n = s.nextInt();
      for(int i=1;i<n+1;i++) edge[i] = new ArrayList();
      for(int i=0;i<n-1;i++)
      {
          int x = s.nextInt();
          int y = s.nextInt();
          edge[x].add(y);
          edge[y].add(x);
      }
      dfs(1);
      
      for(int i=2;i<=n;i++)
      {
          System.out.println(parent[i]);
      }
    }
}

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.kt
import java.util.*

var edge = Array<ArrayList<Int>>(100001){ArrayList()}
var marked = Array<Boolean>(100001){false}
var parent = Array<Int>(100001){0}
var cnt = 0

fun dfs(n : Int)
{
    
    marked[n] = true
    for(x in edge[n])
    {
        if(!marked[x])
        {
            parent[x] = n
            dfs(x)
        }
        
    }
}
fun main() {
    val s = Scanner(System.`in`);
    var n = s.nextInt()
    
    for(i in 0..n-2)
    {
        val x = s.nextInt()
        val y = s.nextInt()
        edge[x].add(y)
        edge[y].add(x)
    }
    dfs(1)
    val sb = StringBuilder()
    for(i in 2..n)
    {
        sb.append(parent[i])
        sb.append('\n')
    }
    print(sb)
}